geetools/ee_image_collection.py

```diff
@@ -1740,3 +1740,113 @@ def to_int(d):
+            reducer: THe reducer to apply.
+            idPropertyType: The type of the idProperty. Default is ee.Number. As Dates are stored as numbers in metadata, we need to know what parsing to apply to the property in advance.
+            dateFormat: If a date format is used for the IdProperty, the values will be formatted as "YYYY-MM-ddThh-mm-ss". You can specify any other format compatible with band names.
+            numberFormat: If a number format is used for the IdProperty, the values will be formatted as a string  ("%s"). You can specify any other format compatible with band names.
+            geometry: The region over which to reduce the data.
+            scale: A nominal scale in meters to work in.
+            crs: The projection to work in. If unspecified, the projection of the image's first band is used. If specified in addition to scale, rescaled to the specified scale.
+            crstransform: The list of CRS transform values. This is a row-major ordering of the 3x2 transform matrix. This option is mutually exclusive with 'scale', and replaces any transform already set on the projection.
+            bestEffort: If the polygon would contain too many pixels at the given scale, compute and use a larger scale which would allow the operation to succeed.
+            maxPixels: The maximum number of pixels to reduce.
+            tileScale: A scaling factor between 0.1 and 16 used to adjust aggregation tile size; setting a larger tileScale (e.g., 2 or 4) uses smaller tiles and may enable computations that run out of memory with the default.
+
+        Returns:
+            A dictionary with the reduced values for each image.
+
+        Examples:
+            .. code-block:: python
+
+                import ee, geetools
+
+                ee.Initialize()
+
+                collection = (
+                    ee.ImageCollection("LANDSAT/LC08/C01/T1_TOA")
+                    .filterBounds(ee.Geometry.Point(-122.262, 37.8719))
+                    .filterDate("2014-01-01", "2014-12-31")
+                )
+                data = collection.geetools.reduceRegion("mean", geometry=ee.Geometry.Point(-122.262, 37.8719), scale=30)
+                print(data.getInfo())
+        """
+        # filter the data to the bounds of the region of interest. This will reduce the amount of
+        # data to process and speed up the computation. We also need to mosaic together images that
+        # have the same idProperty to avoid conflicts.
+        ic = self._obj.filterBounds(geometry)
+        propertyList = ic.aggregate_array(idProperty).distinct()
+        imageList = propertyList.map(lambda p: ic.filter(ee.Filter.eq(idProperty, p)).mosaic())
```

Comment:
This is assuming that the only case in which the `idProperty` is repeated is when images with the same date do not overlap, but are one next to each other, like in Landsat or Sentinel images, so they need to be mosaicked. Ok, this might not cover all cases (for example, if the user passes `dateFormat="DDD"`) but if it's well documented it's ok, the user will need to post-process the results in such case.

To cover cases like my example the function should have a `propertyReducer` param and use this instead of `.mosaic`.

But, if you consider that this creates too many params and the use cases are marginal, then let's continue with this approach.

Reply:
I like the idea of a propertyReducer, it's not so much work to code and easy to document. I guess most people will not change it but at least it will be there.

Reply:
### short answer

ah then I prefer to avoid this complexity. I fear we introduce too many steps and the error become impossible to understand by the user. I prefer to document the `DDD` use case in a documentation example so the user understand that we don't do property aggregation. In fact what we tolerate is reduction when the image is cut into smaller pieces and I don't think we should go any farther. If it's more complex then the `` `imageCollection` needs to be updated before calling the method.

### long answer 
The driven argument for me when building this method is: A user want to use reduceRegion on an imageCollection but it's complicated to map by yourself. Which leads to the following basic workflow: 1 reduction per image

As we exchanged some extra complexity was introduced: 
- 1st complexity: The name of the reduction must be associated to a property and must be converted into a band compatible name it leads to 2 parameters (idProperty and propertyType) and a little if loop. (initially it was dates and that's all)
- 2nd complexity: User should be able to decide the format of such a name (note that there are 0 guarantees they will choose something band compatible) 2 more parameters (dateFormat, numberFormat).
- 3rd complexity: The area of interest covers an area where the property can be the same for multiple image (like raw imagery where each day is splitted into multiple images) 1 more parameter (propertyReducer) and 1 more reduction. 

Now the use case you try cover is: the user selected by himself a naming convention that is bad as it creates duplicate where there was none and we should guess what he really wanted to do. Our current best guess is, he wanted to reduce them earlier than the propertyReducer. The use case you oversee is the DOY representation but is it the only one ? is it really what the user wanted ? 

I would vastly prefer to leave the complexity where it currently sit and adress this precise use case in a documentation example, documentation that can be extended if new needs come up. to get it right the user only need to compute DOY as a metadata of each image and then use `reduceRegion`.

Why do I want to limit complexity ? Because as everything is run server side we have 0 control on the error provided and there is very little chance the user will understand that the method behave unexpectidely because internally we are rebuilding his ImageCollection using a provieded name that requires a time reduction.